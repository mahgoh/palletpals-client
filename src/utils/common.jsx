export function classNames(...classes) {
  return classes.filter(Boolean).join(' ')
}

export function productImageURL(imageId) {
  return `${GLOBAL.API_URL}/product-images/${imageId}`
}

export function parseDateTime(dateTimeString, locale = 'en') {
  const datetime = new Date(dateTimeString)
  const options = {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: 'numeric',
    minute: '2-digit',
  }

  return datetime.toLocaleString(locale, options)
}

export function formatPrice(price) {
  return `CHF ${price.toFixed(2)}`
}

export function formatOrderId(id) {
  return id.toString().padStart(6, '0')
}
