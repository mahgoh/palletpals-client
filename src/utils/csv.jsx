export async function processCSV(file) {
  try {
    const csv = await readCSV(file)
    const parsed = parseCSV(csv)

    return parsed
  } catch (e) {
    throw e
  }
}

function readCSV(file) {
  return new Promise((resolve, reject) => {
    if (file === null) reject('No file selected')

    // Check if the file is an image.
    if (file.type && !file.type === 'text/csv') {
      console.log('File is not csv.', file.type, file)
      reject('File is not csv.')
    }

    const reader = new FileReader()
    reader.addEventListener('load', (event) => {
      resolve(event.target.result)
    })
    reader.readAsText(file)
  })
}

/**
 *
 * Target:
 * {
 *   "km": [10,20,30],
 *   "pallets": [1,2,3],
 *   "price": [[5,6,7],[9,11,13],[15,17,19]]
 * }
 * @param {string} csv
 * @returns
 */

function parseCSV(csv) {
  const lines = csv.split('\n')

  // how entries are separated
  const delimiter = csv.includes(',') ? ',' : ';'

  const head = lines.shift()

  // palletArray is the first row except the first column
  // integers
  const palletArray = head
    .split(delimiter)
    .slice(1)
    .map((p) => parseInt(p))

  // kmArray is the first column except the first row
  // integers
  const kmArray = lines.map(
    (line) => line.split(delimiter).map((km) => parseInt(km))[0]
  )

  // priceMatrix is the rest of the table as two dimensional array
  // floats
  const priceMatrix = lines.map((line) =>
    line
      .split(delimiter)
      .slice(1)
      .map((p) => parseFloat(p))
  )

  return {
    kmArray,
    palletArray,
    priceMatrix,
  }
}
