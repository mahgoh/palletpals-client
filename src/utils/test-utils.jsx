import { cloneElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { I18nextProvider } from 'react-i18next'
import { render } from '@testing-library/react'
import i18n from '../i18n'

import { AuthProvider } from '@/services/auth'
import { AppearanceProvider } from '@/services/appearance'
import { CartProvider } from '@/services/cart'
import { LanguageProvider } from '@/services/language'
import { NotificationProvider } from '@/services/notification'

/**
 * Render component with router wrapper and option for specifying the current route.
 *
 * @param {JSX} component
 * @param {Object} options
 * @returns
 */
const renderWithRouter = (component, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route)

  return {
    ...render(component, { wrapper: BrowserRouter }),
  }
}

/**
 * Render component with router wrapper and i18next provider.
 * @param {JSX} component
 * @param {Object} options
 * @returns
 */
const renderWithRouterWithi18n = (component, options) => {
  const comp = cloneElement(component, {
    changeLanguage: (lng) => {
      i18n.changeLanguage(lng)
      rerender(<I18nextProvider i18n={i18n}>{comp}</I18nextProvider>)
    },
  })
  const defaultRender = renderWithRouter(
    <I18nextProvider i18n={i18n}>{comp}</I18nextProvider>,
    options
  )
  const { rerender } = defaultRender
  return defaultRender
}

const renderComplete = (component, options) => {
  return {
    ...renderWithRouterWithi18n(
      <AuthProvider>
        <AppearanceProvider>
          <LanguageProvider>
            <CartProvider>
              <NotificationProvider>{component}</NotificationProvider>
            </CartProvider>
          </LanguageProvider>
        </AppearanceProvider>
      </AuthProvider>,
      options
    ),
  }
}

/**
 * Render component with specified renderer
 *
 * @param {JSX} component
 * @param {Object} param1
 * @returns
 */
const customRender = (component, { renderer = 'default', ...args } = {}) => {
  switch (renderer) {
    case 'router':
      return renderWithRouter(component, { ...args })
    case 'router-i18n':
      return renderWithRouterWithi18n(component, { ...args })
    case 'complete':
      return renderComplete(component, { ...args })
    case 'default':
    default:
      return {
        ...render(component),
      }
  }
}

export * from '@testing-library/react'
// export { default as userEvent } from '@testing-library/user-event'
export { customRender as render }
