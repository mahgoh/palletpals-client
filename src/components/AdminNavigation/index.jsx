import { NavLink } from 'react-router-dom'
import { useTranslation } from 'react-i18next'
import { classNames } from '@/utils/common'

export default function AdminNavigation() {
  const { t } = useTranslation()

  function renderNavLink(path, label) {
    return (
      <NavLink
        key={path}
        to={path}
        className={({ isActive }) =>
          classNames(
            'mb-2 w-full -translate-x-3 rounded-md py-1 px-3 transition-colors last:mb-0 hover:bg-orange-500/10  hover:text-orange-600 focus:text-orange-600 sm:mr-2 sm:mb-0 sm:w-auto sm:last:mr-0',
            isActive ? 'bg-transparent font-bold text-orange-600' : ''
          )
        }
      >
        {t(label)}
      </NavLink>
    )
  }
  function renderNavLinks() {
    const links = [
      {
        path: '/admin/products',
        label: t('common.product.title', { numProducts: 2 }),
      },
      {
        path: '/admin/service-providers',
        label: t('common.service-provider.title', { numServiceProviders: 2 }),
      },
      {
        path: '/admin/warehouses',
        label: t('common.warehouse.title', { numWarehouses: 2 }),
      },
    ]

    return links.map((link) => renderNavLink(link.path, link.label))
  }

  return (
    <div className="border-b border-gray-200 py-4 dark:border-gray-700">
      <div className="mx-auto flex flex-col items-center px-4 sm:flex-row lg:max-w-6xl lg:px-10">
        {renderNavLinks()}
      </div>
    </div>
  )
}
