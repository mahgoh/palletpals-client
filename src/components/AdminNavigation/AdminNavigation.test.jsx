import React from 'react'
import { render, screen } from '@/utils/test-utils'
import AdminNavigation from '.'

describe('admin navigation', () => {
  it('should render with text content', () => {
    const { container } = render(<AdminNavigation />, { renderer: 'complete' })
    expect(container).toMatchSnapshot()
  })

  it('should not render with children', () => {
    const { container } = render(
      <AdminNavigation>
        <div>should not be rendered</div>
      </AdminNavigation>,
      { renderer: 'complete' }
    )

    expect(screen.queryByText('should not be rendered')).toBeNull()
    expect(container).toMatchSnapshot()
  })
})
