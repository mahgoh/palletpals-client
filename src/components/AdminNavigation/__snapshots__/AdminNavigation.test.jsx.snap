// Vitest Snapshot v1

exports[`admin navigation > should not render with children 1`] = `
<div>
  <div
    class="border-b border-gray-200 py-4 dark:border-gray-700"
  >
    <div
      class="mx-auto flex flex-col items-center px-4 sm:flex-row lg:max-w-6xl lg:px-10"
    >
      <a
        class="mb-2 w-full -translate-x-3 rounded-md py-1 px-3 transition-colors last:mb-0 hover:bg-orange-500/10  hover:text-orange-600 focus:text-orange-600 sm:mr-2 sm:mb-0 sm:w-auto sm:last:mr-0"
        href="/admin/products"
      >
        Products
      </a>
      <a
        class="mb-2 w-full -translate-x-3 rounded-md py-1 px-3 transition-colors last:mb-0 hover:bg-orange-500/10  hover:text-orange-600 focus:text-orange-600 sm:mr-2 sm:mb-0 sm:w-auto sm:last:mr-0"
        href="/admin/service-providers"
      >
        Service Providers
      </a>
      <a
        class="mb-2 w-full -translate-x-3 rounded-md py-1 px-3 transition-colors last:mb-0 hover:bg-orange-500/10  hover:text-orange-600 focus:text-orange-600 sm:mr-2 sm:mb-0 sm:w-auto sm:last:mr-0"
        href="/admin/warehouses"
      >
        Warehouses
      </a>
    </div>
  </div>
</div>
`;

exports[`admin navigation > should render with text content 1`] = `
<div>
  <div
    class="border-b border-gray-200 py-4 dark:border-gray-700"
  >
    <div
      class="mx-auto flex flex-col items-center px-4 sm:flex-row lg:max-w-6xl lg:px-10"
    >
      <a
        class="mb-2 w-full -translate-x-3 rounded-md py-1 px-3 transition-colors last:mb-0 hover:bg-orange-500/10  hover:text-orange-600 focus:text-orange-600 sm:mr-2 sm:mb-0 sm:w-auto sm:last:mr-0"
        href="/admin/products"
      >
        Products
      </a>
      <a
        class="mb-2 w-full -translate-x-3 rounded-md py-1 px-3 transition-colors last:mb-0 hover:bg-orange-500/10  hover:text-orange-600 focus:text-orange-600 sm:mr-2 sm:mb-0 sm:w-auto sm:last:mr-0"
        href="/admin/service-providers"
      >
        Service Providers
      </a>
      <a
        class="mb-2 w-full -translate-x-3 rounded-md py-1 px-3 transition-colors last:mb-0 hover:bg-orange-500/10  hover:text-orange-600 focus:text-orange-600 sm:mr-2 sm:mb-0 sm:w-auto sm:last:mr-0"
        href="/admin/warehouses"
      >
        Warehouses
      </a>
    </div>
  </div>
</div>
`;
